crypto/multisig/threshold_multisig.go

```diff
@@ -0,0 +1,78 @@
+package multisig
+
+import (
+	"github.com/tendermint/tendermint/crypto"
+	"github.com/tendermint/tendermint/crypto/tmhash"
+)
+
+// ThresholdMultiSignaturePubKey implements a K of N threshold multisig.
+type ThresholdMultiSignaturePubKey struct {
+	K       uint            `json:"threshold"`
+	Pubkeys []crypto.PubKey `json:"pubkeys"`
+}
+
+var _ crypto.PubKey = &ThresholdMultiSignaturePubKey{}
+
+// NewThresholdMultiSignaturePubKey returns a new ThresholdMultiSignaturePubKey.
+func NewThresholdMultiSignaturePubKey(k int, pubkeys []crypto.PubKey) crypto.PubKey {
+	if len(pubkeys) < k {
+		panic("threshold k of n multisignature: len(pubkeys) < k")
+	}
+	return &ThresholdMultiSignaturePubKey{uint(k), pubkeys}
+}
+
+// VerifyBytes expects sig to be an amino encoded version of a MultiSignature.
+// Returns true iff the multisignature contains k or more signatures
+// for the correct corresponding keys,
+// and all signatures are valid. (Not just k of the signatures)
+// The multisig uses a bitarray, so multiple signatures for the same key is not
+// a concern.
+func (pk *ThresholdMultiSignaturePubKey) VerifyBytes(msg []byte, marshalledSig []byte) bool {
+	var sig *Multisignature
+	err := cdc.UnmarshalBinary(marshalledSig, &sig)
+	if err != nil {
+		return false
+	}
+	size := sig.BitArray.Size()
+	if len(sig.Sigs) < int(pk.K) || len(pk.Pubkeys) != size || sig.BitArray.NumOfTrueBitsBefore(size) < int(pk.K) {
```

Comment:
Should we also do a consistency check that `len(sig.Sigs) < size`?

Maybe it would be better if we split this up and evaluate the concerns separately.

Like first we check the size of the bitarray against the number of sigs and the number of pubkeys.

Then we handle the K stuff.

Reply:
but we allow `len(sigs.Sigs) > k`, so we should set an upper bound.

Reply:
Oh woops, your totally right, great catch!